### PR TITLE
Add the cluster version to enrich policies when they are created.

### PR DESCRIPTION
**Pull request by jbaiera**
This PR adds version information to an EnrichPolicy when it is created to aid with backwards compatibility logic. The old EnrichPolicy object has been renamed to `EnrichPolicyDefinition` and a new `EnrichPolicy` object has been created, one that holds the version of Elasticsearch that the policy was created on. I also added the name to the policy object since in many places we are passing the name around along with the policy definition. All response objects have been updated to return the new policy structure. Now when a get operation is performed, the following is returned:

```json
{
  "name": "policy-name",
  "version_created": "8.0.0",
  "definition": {
    "type": "exact_match",
    "indices": ["source_index"],
    "query": { ... },
    "enrich_key": "keyname",
    "enrich_values": ["field1", "field2", "field3"]
  }
}
```

### CODE REVIEW

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/enrich/EnrichPolicy.java [hunk 1]
```diff
@@ -5,142 +5,93 @@
...
-        parser.declareObject(ConstructingObjectParser.optionalConstructorArg(), (p, c) -> {
-            XContentBuilder contentBuilder = XContentBuilder.builder(p.contentType().xContent());
-            contentBuilder.generator().copyCurrentStructure(p);
-            return new QuerySource(BytesReference.bytes(contentBuilder), contentBuilder.contentType());
-        }, QUERY);
-        parser.declareStringArray(ConstructingObjectParser.constructorArg(), INDICES);
-        parser.declareString(ConstructingObjectParser.constructorArg(), ENRICH_KEY);
-        parser.declareStringArray(ConstructingObjectParser.constructorArg(), ENRICH_VALUES);
+        PARSER.declareString(ConstructingObjectParser.constructorArg(), NAME);
+        PARSER.declareField(ConstructingObjectParser.constructorArg(), (p, c) -> Version.fromString(p.text()), VERSION_CREATED,
```
Comment (jakelandis) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/enrich/EnrichPolicy.java]: Can newer ES versions fromString versions that are not explicitly defined. e.g. what happens if version 6.8.0 is the version serialized can this read it ?

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/enrich/action/ListEnrichPolicyAction.java [hunk 2]
```diff
@@ -47,17 +47,19 @@ public ActionRequestValidationException validate() {
 
     public static class Response extends ActionResponse implements ToXContentObject {
 
-        private final List<EnrichPolicy.NamedPolicy> policies;
+        private final List<EnrichPolicy> policies;
 
         public Response(Map<String, EnrichPolicy> policies) {
             Objects.requireNonNull(policies, "policies cannot be null");
             // use a treemap to guarantee ordering in the set, then transform it to the list of named policies
```
Comment (jakelandis) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/enrich/action/ListEnrichPolicyAction.java]: nit: remove the comment, this no longer uses a Treemap

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/enrich/EnrichPolicy.java [hunk 3]
```diff
@@ -5,142 +5,93 @@
...
-        if (query != null) {
-            builder.field(QUERY.getPreferredName(), query.getQueryAsMap());
-        }
-        builder.array(INDICES.getPreferredName(), indices.toArray(new String[0]));
-        builder.field(ENRICH_KEY.getPreferredName(), enrichKey);
-        builder.array(ENRICH_VALUES.getPreferredName(), enrichValues.toArray(new String[0]));
+        builder.field(NAME.getPreferredName(), name);
+        builder.field(VERSION_CREATED.getPreferredName());
+        versionCreated.toXContent(builder, params);
+        builder.startObject(DEFINITION.getPreferredName());
```
Comment (jakelandis) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/enrich/EnrichPolicy.java]: does the definition need be a sub object ?

File: x-pack/plugin/enrich/qa/rest/src/test/resources/rest-api-spec/test/enrich/10_basic.yml [hunk 4]
```diff
@@ -19,19 +19,20 @@
   - do:
       enrich.get_policy:
         name: policy-crud
-  - match: { type: exact_match }
-  - match: { indices: ["bar*"] }
-  - match: { enrich_key: baz }
-  - match: { enrich_values: ["a", "b"] }
+  - match: { name: policy-crud }
+  - match: { definition.type: exact_match }
```
Comment (jakelandis) [inline x-pack/plugin/enrich/qa/rest/src/test/resources/rest-api-spec/test/enrich/10_basic.yml]: asked above ^^ if can avoid the `definition` sub object .

**Comment by jakelandis** [comment]
Is it possible to keep this as single policy object which can optionally serialize/deserialize the extra two fields ?

**Comment by jbaiera** [comment]
> Is it possible to keep this as single policy object which can optionally serialize/deserialize the extra two fields ?

I wanted to avoid having fields on the policy object that would need to be handled specially for the purposes of serialization. More specifically, a user shouldn't be able to set the `version_created` field on a policy, but the policy would still need to be able to read/write the field as JSON for the purpose of existing in the cluster state. I wasn't sure if there was a way of saying "this field isn't allowed to be set by the user, but should be allowed to be read from the cluster state" without some awkward code along side it with the caveats mentioned. The separated object just felt a little cleaner.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/enrich/EnrichPolicy.java (hunk 1 above)
Comment (jbaiera) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/enrich/EnrichPolicy.java]: Yes - The `fromString()` method condenses the version string given down into the integral version id. If the version id isn't defined in the known list of versions, it is still created, but the version of lucene that it associates with the returned version is a guess. I don't foresee this being an issue, but I could be off base here.

**Comment by martijnvg** [comment]
Maybe introduce a second map in `EnrichMetadata` that contains a new class with the index version instead of a new class that wraps a policy?

**Comment by jbaiera** [comment]
Closing this out in favor of #45021 - It accomplishes the same thing, but with less ripple effect.

**Pull request closed without merging**